### PR TITLE
Fix usage of deprecated methods in controllers

### PR DESCRIPTION
**Pull request by ZRazor**
I am targeting this branch, because it's BC.

## Changelog

```markdown
### Fixed
- Fixed usage of deprecated methods in controllers (Improve compatibility with SF 3)
```

### CODE REVIEW

File: src/ProductBundle/Controller/ProductVariationAdminController.php [hunk 1]
```diff
@@ -36,8 +41,15 @@ public function createAction(Request $request = null)
             throw new AccessDeniedException();
         }
 
+        // NEXT_MAJOR: Keep FQCN when bumping Symfony requirement to 2.8+.
+        if (method_exists('Symfony\Component\Form\AbstractType', 'getBlockPrefix')) {
```
Comment (greg0ire) [inline src/ProductBundle/Controller/ProductVariationAdminController.php]: Please use the `::class` notation as much as you can.

Comment (ZRazor) [reply src/ProductBundle/Controller/ProductVariationAdminController.php]: Is it better now?

File: src/ProductBundle/Controller/ProductVariationAdminController.php [hunk 2]
```diff
@@ -36,8 +42,15 @@ public function createAction(Request $request = null)
             throw new AccessDeniedException();
         }
 
+        // NEXT_MAJOR: Keep FQCN when bumping Symfony requirement to 2.8+.
+        if (method_exists(\Symfony\Component\Form\AbstractType::class, 'getBlockPrefix')) {
```
Comment (greg0ire) [inline src/ProductBundle/Controller/ProductVariationAdminController.php]: It is better, but please use a use statement too

Comment (ZRazor) [reply src/ProductBundle/Controller/ProductVariationAdminController.php]: Done

**Comment by greg0ire** [comment]
Please write a changelog directed at end users. Keep in mind that they probably do not care what file or class you changed, but what features you provided / what bugs you fixed. Try to answer the question "What benefit do I get from upgrading?"
More on http://keepachangelog.com

Also, in your case, end users do not care about phpdoc changes

File: src/BasketBundle/Controller/BasketController.php [hunk 3]
```diff
@@ -522,7 +524,19 @@ public function finalReviewStepAction()
...
+            'tac_error' => 'POST' == $this->getRequest()->getMethod(),
         ]);
     }
+
+    /**
+     * @return Request
+     */
+    public function getRequest()
+    {
+        if ($this->container->has('request_stack')) {
```
Comment (jordisala1991) [inline src/BasketBundle/Controller/BasketController.php]: I think we can safely rewrite those to always return `return $this->container->get('request_stack')->getCurrentRequest();`

File: src/BasketBundle/Controller/BasketController.php [hunk 4]
```diff
@@ -522,7 +524,19 @@ public function finalReviewStepAction()
...
             'basket' => $basket,
-            'tac_error' => 'POST' == $this->get('request')->getMethod(),
+            'tac_error' => 'POST' == $this->getRequest()->getMethod(),
         ]);
     }
+
+    /**
+     * @return Request
+     */
+    public function getRequest()
```
Comment (jordisala1991) [inline src/BasketBundle/Controller/BasketController.php]: Can't we make those private?

File: src/ProductBundle/Controller/ProductVariationAdminController.php [hunk 5]
```diff
@@ -36,8 +43,15 @@ public function createAction(Request $request = null)
...
+        // NEXT_MAJOR: Keep FQCN when bumping Symfony requirement to 2.8+.
+        if (method_exists(AbstractType::class, 'getBlockPrefix')) {
+            $integerType = IntegerType::class;
+        } else {
+            $integerType = 'integer';
+        }
+
         $form = $this->createFormBuilder(null, [])
-            ->add('number', 'integer', [
+            ->add('number', $integerType, [
```
Comment (jordisala1991) [inline src/ProductBundle/Controller/ProductVariationAdminController.php]: Remove the `if` and make this: `IntegerType::class`

File: src/BasketBundle/Controller/BasketController.php (hunk 4 above)
Comment (ZRazor) [reply src/BasketBundle/Controller/BasketController.php]: No, this method is public in parent class in SF 2.8. But we can rename it, make private and replace all usages of getRequest().

File: src/ProductBundle/Controller/ProductVariationAdminController.php (hunk 5 above)
Comment (ZRazor) [reply src/ProductBundle/Controller/ProductVariationAdminController.php]: Okay, SF < 2.8 is dropped, so no more this stuff with `if` anywhere ;)

Comment (jordisala1991) [reply src/ProductBundle/Controller/ProductVariationAdminController.php]: I am preparing a PR to fully replace those, If you want, don't change this one

File: src/BasketBundle/Controller/BasketController.php (hunk 4 above)
Comment (jordisala1991) [reply src/BasketBundle/Controller/BasketController.php]: Not sure if it is a good idea to name it like the parent, because on SF 3.4 I think it is removed or final?

Maybe give it another name like: `getCurrentRequest`?

**Comment by jordisala1991** [comment]
Maybe the changelog could be something like, Improve compatibility with SF 3?

File: src/BasketBundle/Controller/BasketController.php (hunk 3 above)
Comment (ZRazor) [reply src/BasketBundle/Controller/BasketController.php]: Done

File: src/BasketBundle/Controller/BasketController.php (hunk 4 above)
Comment (ZRazor) [reply src/BasketBundle/Controller/BasketController.php]: Done

File: src/BasketBundle/Controller/BasketController.php [hunk 6]
```diff
@@ -47,7 +49,7 @@ public function indexAction($form = null)
         ]);
 
         // always validate the basket
-        if (!$form->isBound()) {
+        if (!$form->isSubmitted()) {
```
Comment (jordisala1991) [inline src/BasketBundle/Controller/BasketController.php]: This change is not related to this PR, is this because of Sf3 compat?

Comment (ZRazor) [reply src/BasketBundle/Controller/BasketController.php]: I have already fixed these deprecations in previous PR, but missed this one. And yes - this is for SF 3.

**Comment by jordisala1991** [comment]
Can you simplify some calls? There are cases where `getCurrentRequest` gets called like 5 or 6 times, would be better to store `$request` in a variable and use it everywhere.

**Comment by ZRazor** [comment]
@jordisala1991 Done.

File: src/ProductBundle/Controller/ProductController.php [hunk 7]
```diff
@@ -276,4 +279,12 @@ public function variationToProductAction($productId, $slug)
 
         return $response;
     }
+
+    /**
+     * @return Request
+     */
+    private function getCurrentRequest()
```
Comment (OskarStark) [inline src/ProductBundle/Controller/ProductController.php]: can we add this to the CRUDController? Or create an BaseEcommerceCRUDController to avoid this duplication?

Comment (ZRazor) [reply src/ProductBundle/Controller/ProductController.php]: mb use trait?

Comment (OskarStark) [reply src/ProductBundle/Controller/ProductController.php]: sounds good 👍 

WDYT @greg0ire ?

Comment (greg0ire) [reply src/ProductBundle/Controller/ProductController.php]: It's something that will be removed so a trait kinda makes sense here, because it's arguably easier to remove, and allows to keep that method private.

Comment (OskarStark) [reply src/ProductBundle/Controller/ProductController.php]: we should add a `NEXT_MAJOR` above the usage.

What about a deprecation in the trait when it will be removed?

Comment (ZRazor) [reply src/ProductBundle/Controller/ProductController.php]: I made a try.
We'll remove `CurrentRequestTrait` in 3.x and inject Request $request in actions parameters.

**Comment by greg0ire** [comment]
Please rebase to get rid of the merge commit.

**Comment by ZRazor** [comment]
@greg0ire Done

File: src/Component/Controller/CurrentRequestTrait.php [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+ * (c) Thomas Rabaix <thomas.rabaix@sonata-project.org>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+namespace Sonata\Component\Controller;
+
+use Symfony\Component\HttpFoundation\Request;
+/**
+ * NEXT_MAJOR: Remove this class
```
Comment (greg0ire) [inline src/Component/Controller/CurrentRequestTrait.php]: It's not a class

File: src/Component/Controller/CurrentRequestTrait.php [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+ *
+ * @deprecated Class to be removed in 3.x
```
Comment (greg0ire) [inline src/Component/Controller/CurrentRequestTrait.php]: Still not a class :P

Comment (ZRazor) [reply src/Component/Controller/CurrentRequestTrait.php]: My bad ;)

File: src/Component/Controller/CurrentRequestTrait.php [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+
+use Symfony\Component\HttpFoundation\Request;
+/**
+ * NEXT_MAJOR: Remove this trait
+ *
+ * @deprecated Trait to be removed in 3.x
+ *
+ * @author Anton Zlotnikov <exp.razor@gmail.com>
+ */
+trait CurrentRequestTrait
```
Comment (core23) [inline src/Component/Controller/CurrentRequestTrait.php]: 👎 for this trait, because it could be used by other people => we couldn't drop this on a stable branch

**Comment by jordisala1991** [comment]
Can we get an agreement here? to not make @ZRazor implement something and then go back and so on?

Do we need a trait here? IMO it reduces some code needed, but I agree on the fact that I prefer to add code that we can easily remove if we need to, and not be forced to keep it because BC compat.

My opinion is to go back to private methods. WDYT?

**Comment by ZRazor** [comment]
So, trait or methods in each controller?
Trait is deprecated, so we warn users from the beginning that it will be removed. Anyway we'll soon inject Request into actions params, and there is no big difference between removing a method at the end of 6 controllers or removing trait and it use statements.

**Comment by ZRazor** [comment]
@core23 Done

**Comment by jordisala1991** [comment]
Can you do a rebase? we merged a PR to improve our lint process

**Comment by ZRazor** [comment]
Oh this rebasing without merge commit 🌚 
@jordisala1991, Done.

**Comment by greg0ire** [comment]
Thanks @ZRazor !
